externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala

```diff
@@ -103,19 +104,20 @@ object SparkDatasetHelper extends Logging {
       timestampAsString: Boolean): DataFrame = {
 
     val quotedCol = (name: String) => col(quoteIfNeeded(name))
+    val tf = getTimeFormatters
```

Comment:
I'm not sure if `HiveResult.getTimeFormatters` is thread safe and serializable, `convertTopLevelComplexTypeToHiveString` method is only for handling complex types in arrow mode, @cfmcgrady do you have any suggestions?

Reply:
Shall we roll back to create TimeFormatters inside `convertTopLevelComplexTypeToHiveString`, if `HiveResult.getTimeFormatters` is probably not thread-safe ?

Reply:
> Shall we roll back to create TimeFormatters inside `convertTopLevelComplexTypeToHiveString`, if `HiveResult.getTimeFormatters` is probably not thread-safe ?

In that case, we could just do an adjustment like this

```
 RowSet.toHiveString((row, st), nested = true, timeFormatters = getTimeFormatters)
```

Reply:
> In that case, we could just do an adjustment like this
> ```
>  RowSet.toHiveString((row, st), nested = true, timeFormatters = getTimeFormatters)
> ```

the partition/task level LGTM.

Reply:
It may not be thread-safe, so we should try to control its usage scope.

FYI: https://github.com/apache/spark/blob/43ca0b929ab3c2f10d1879e5df622195564f8885/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala#L198

https://github.com/apache/spark/blob/43ca0b929ab3c2f10d1879e5df622195564f8885/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala#L532